### ISSUE TITLE
Support for transactions

### ISSUE THREAD

**Issue by amirouche** [twmht/python-rocksdb#36]
It seems the bindings are missing transaction support.

**Comment by twmht** [comment]
Thank you for using python-rocksdb. these features are coming soon since I am re-write whole project to make it more easily to maintain.

**Comment by amirouche** [comment]
tx!

**Comment by NightTsarina** [comment]
It has been 2 years since this issue was opened. Any news on this? I was considering using rocksdb for a project written in python, but not having transactions is a showstopper!

**Comment by iFA88** [comment]
You can use WriteBatch for that, or even a custom writebatch which stores the key data and statuses for from that.

**Comment by NightTsarina** [comment]
Not really, because I need read-write conflict detection by using GetForUpdate.

**Comment by iFA88** [comment]
> Not really, because I need read-write conflict detection by using GetForUpdate.

Yeah, you can do that in python. You can use `bisect` for automatic ordering when you need iter too.

**Comment by NightTsarina** [comment]
> > Not really, because I need read-write conflict detection by using GetForUpdate.
> 
> Yeah, you can do that in python. You can use `bisect` for automatic ordering when you need iter too.

I am not sure I understand what you mean.. How does bisect help me with this?

**Comment by iFA88** [comment]
> I am not sure I understand what you mean.. How does bisect help me with this?

So simple:
```
class OwnWriteBatch:
  def __init__(self, db):
    self.db = db
    self.keys = {}
  def get(self, key, *args, **kwargs):
    if key not in self.keys:
      return self.db.get(key, *args, **kwargs)
    return self.keys[key][1] if self.keys[key][0] else None
  def put(self, key, value):
     self.keys[key] = [True, value]
  def delete(self, key):
     self.keys[key] = [False, None]
  def write(self):
     wb = rocksdb.WriteBatch()
     for k, (s, v) in self.keys:
       if s:
         wb.put(k, v)
       else:
         wb.delete(k)
    self.db.write(wb)
```

**Comment by NightTsarina** [comment]
Right. But this still does not provide read-write or write-write conflict detection. I *do* need transactions for that.

**Comment by iFA88** [comment]
> Right. But this still does not provide read-write or write-write conflict detection. I _do_ need transactions for that.

Can you explain that a little more?

**Comment by NightTsarina** [comment]
Transactions can give you guarantees that no other thread has updated an entry your transaction is modifying or depending upon. See https://github.com/facebook/rocksdb/wiki/Transactions

**Comment by iFA88** [comment]
I'm using `threading.RLock` for that. I think what you want there is a "simple" workaround for that, but the `Transactions` is currently not supported.

**Comment by NightTsarina** [comment]
I appreciate you trying to help, but you are not understanding the problem and I am not looking for workarounds. This bug is about the missing support for transactions as provided by rocksdb.

**Comment by amirouche** [comment]
for what it is worth sqlite lsm extension support transactions and you build it standalone as shared lib.

HTH.

**Comment by amirouche** [comment]
https://github.com/coleifer/python-lsm-db
